### PR TITLE
v4l2m2m deinterlace support

### PR DESCRIPTION
**Pull request by popcornmix**
This provides access to the mmal based image_fx component to enable deinterlace through v4l2m2m
with a suitable ffmmpeg. e.g. https://github.com/popcornmix/FFmpeg/tree/deinterlace

### CODE REVIEW

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c [hunk 1]
```diff
@@ -75,6 +75,11 @@ static unsigned int debug;
 module_param(debug, uint, 0644);
 MODULE_PARM_DESC(debug, "activates debug info (0-3)");
 
+static bool advanced = true;
+module_param(advanced, bool, 0644);
+MODULE_PARM_DESC(advanced, "Use advanced deinterlace");
```
Comment (6by9) [inline drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c]: ```advanced``` is a bit generic as module parameters go.

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c [hunk 2]
```diff
@@ -75,6 +75,11 @@ static unsigned int debug;
...
+
+
```
Comment (6by9) [inline drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c:85]: checkpatch will complain of double blank lines.

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c [hunk 3]
```diff
@@ -2360,10 +2365,18 @@ static int bcm2835_codec_create_component(struct bcm2835_codec_ctx *ctx)
...
+			advanced && ctx->q_data[V4L2_M2M_SRC].crop_width <= 800 ?
+			MMAL_PARAM_IMAGEFX_DEINTERLACE_ADV :
+			MMAL_PARAM_IMAGEFX_DEINTERLACE_FAST;
 		struct mmal_parameter_imagefx_parameters params = {
-			.effect = MMAL_PARAM_IMAGEFX_DEINTERLACE_ADV,
-			.num_effect_params = 0,
-			.effect_parameter = { 0 },
+			.effect = effect,
+			.num_effect_params = 4,
+			.effect_parameter = { frame_type,default_frame_interval,half_framerate,use_qpus },
```
Comment (6by9) [inline drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c]: Formatting. A minimum of spaces after each comma, or preferably one per line.

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c [hunk 4]
```diff
@@ -79,6 +79,10 @@ static bool advanced = true;
 module_param(advanced, bool, 0644);
 MODULE_PARM_DESC(advanced, "Use advanced deinterlace");
 
+static int field;
```
Comment (6by9) [inline drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c]: Too generic a name to be a module parameter, and likely to get confused with a local var called ```field```.

```field_override```?

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c [hunk 5]
```diff
@@ -1715,7 +1716,38 @@ static int vidioc_g_selection(struct file *file, void *priv,
 		}
 		break;
 	case ISP:
+		break;
 	case DEINTERLACE:
+		switch (s->target) {
```
Comment (6by9) [inline drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c]: These need some extra validation. COMPOSE targets are only valid on V4L2_BUF_TYPE_VIDEO_CAPTURE, and CROP targets are only valid on V4L2_BUF_TYPE_VIDEO_OUTPUT

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c [hunk 6]
```diff
@@ -1791,7 +1823,32 @@ static int vidioc_s_selection(struct file *file, void *priv,
...
```
Comment (6by9) [inline drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c]: These need some extra validation. COMPOSE targets are only valid on V4L2_BUF_TYPE_VIDEO_CAPTURE, and CROP targets are only valid on V4L2_BUF_TYPE_VIDEO_OUTPUT

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c [hunk 7]
```diff
@@ -2360,10 +2365,18 @@ static int bcm2835_codec_create_component(struct bcm2835_codec_ctx *ctx)
 					      sizeof(enable));
 	} else if (dev->role == DEINTERLACE) {
 		/* Select the default deinterlace algorithm. */
+		int half_framerate = 0;
+		int default_frame_interval = -1; // don't interpolate
```
Comment (6by9) [inline drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c]: /* */ for comments

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c [hunk 8]
```diff
@@ -1522,8 +1522,9 @@ static int vidioc_s_fmt(struct bcm2835_codec_ctx *ctx, struct v4l2_format *f,
 	q_data->fmt = find_format(f, ctx->dev,
 				  f->type == V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE);
 	q_data->crop_width = f->fmt.pix_mp.width;
+	q_data->crop_height = f->fmt.pix_mp.height;
```
Comment (6by9) [inline drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c]: This changes behaviour for the other roles.

This one does feel weird as to why it is required - crop_height should come through s_selection, so why is it needed here?

**Review by 6by9 — COMMENTED** [review]
Patches 2-5 can be squashed in my view - they're all just adding a "fully featured" deinterlace implementation.

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c [hunk 9]
```diff
@@ -2335,6 +2431,29 @@ static int bcm2835_codec_create_component(struct bcm2835_codec_ctx *ctx)
...
+		enum mmal_parameter_imagefx effect =
+			advanced_deinterlace && ctx->q_data[V4L2_M2M_SRC].crop_width <= 800 ?
+			MMAL_PARAM_IMAGEFX_DEINTERLACE_ADV :
+			MMAL_PARAM_IMAGEFX_DEINTERLACE_FAST;
+		struct mmal_parameter_imagefx_parameters params = {
+			.effect = effect,
+			.num_effect_params = 4,
+			.effect_parameter = { frame_type,
+					      default_frame_interval,
+					      half_framerate,use_qpus },
```
Comment (6by9) [inline drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c]: Space after comma, or an extra carriage return.

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c (hunk 6 above)
Comment (6by9) [reply drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c]: Whilst fixed for g_selection, this still applies for s_selection

File: drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c (hunk 8 above)
Comment (6by9) [reply drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c]: Can you just check whether this and the change 2 lines later are actually still needed? Now we're calling s_selection on the correct buffer type they feel unnecessary.

**Review by 6by9 — COMMENTED** [review]
Drop my Signed-off-by from the commit text. Co-developed-by if you're fussed (before your SoB), but seeing as we're both working for the same company it has limited benefit.

**Comment by popcornmix** [comment]
Updated to address comments.

**Comment by 6by9** [comment]
LGTM

**Pull request merged**